### PR TITLE
A0-2733: loop-select-ng

### PR DESCRIPTION
**Pull request by fixxxedpoint**
# Description

Few changes around `loop-select` statements, ensuring no channel is called after `is_terminated` returns `true`, etc.
Introduced `Bottom` type for statically checked infinite computations.

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- mainly refactor

### CODE REVIEW

File: finality-aleph/src/sync/service.rs [hunk 1]
```diff
@@ -730,54 +731,61 @@ where
     }
 
     /// Stay synchronized.
-    pub async fn run(mut self) {
+    pub async fn run(mut self) -> Result<Bottom, String> {
+        if self.additional_justifications_from_user.is_terminated() {
+            return Err(
+                "Channel with additional justifications from user closed before we started.",
+            )?;
```
Comment (ggawryal) [inline finality-aleph/src/sync/service.rs:738]:
```suggestion
            );
```

File: finality-aleph/src/sync/service.rs [hunk 2]
```diff
@@ -730,54 +731,61 @@ where
...
+        }
+        if self.blocks_from_creator.is_terminated() {
+            return Err("Channel with own blocks closed before we started.")?;
```
Comment (ggawryal) [inline finality-aleph/src/sync/service.rs:741]:
```suggestion
            return Err("Channel with own blocks closed before we started.");
```

File: finality-aleph/src/data_io/data_provider.rs [hunk 3]
```diff
@@ -280,25 +280,23 @@ where
...
+            best_block_in_session = self.get_best_block_in_session(best_block_in_session).await;
+            if let Some(best_block) = &best_block_in_session {
+                self.update_data(best_block);
             }
         }
     }
+
+    pub async fn run(mut self, exit: oneshot::Receiver<()>) {
+        tokio::select! {
+            _ = self.main_loop() => error!(target: "aleph-data-store", "Task for refreshing best chain finished."),
```
Comment (ggawryal) [inline finality-aleph/src/data_io/data_provider.rs:296]: Compiler is not able to deduce that this is unreachable? If it isn't, I think `error` is indeed the best option out of `panic!` and `{}`.

**Review by ggawryal — APPROVED** [review]
LGTM, I like fusing notification streams in substrate network wrapper :)

File: finality-aleph/src/block/substrate/status_notifier.rs [hunk 4]
```diff
@@ -64,6 +67,7 @@ impl ChainStatusNotifier<Header> for SubstrateChainStatusNotifier {
                 .map(|block| ChainStatusNotification::BlockImported(block.header))
                 .ok_or(Error::ImportStreamClosed)
             }
+            complete => Err(Error::AllStreamsTerminated)
```
Comment (timorleph) [inline finality-aleph/src/block/substrate/status_notifier.rs:70]: Is `complete` a keyword, or is something that I don't understand going on here?

File: finality-aleph/src/data_io/data_store.rs [hunk 5]
```diff
@@ -237,26 +238,36 @@ where
         )
     }
 
-    pub async fn run(&mut self, mut exit: oneshot::Receiver<()>) {
+    pub async fn run(&mut self, mut exit: oneshot::Receiver<()>) -> Result<(), String> {
```
Comment (timorleph) [inline finality-aleph/src/data_io/data_store.rs:241]: Error type plx.

File: finality-aleph/src/network/gossip/service.rs [hunk 6]
```diff
@@ -503,36 +504,26 @@ impl<N: RawNetwork, ES: EventStream<N::PeerId>, AD: Data, BSD: Data> Service<N,
         info!(target: LOG_TARGET, "{}", status);
     }
 
-    pub async fn run(mut self) {
+    pub async fn run(mut self) -> Result<Bottom, String> {
```
Comment (timorleph) [inline finality-aleph/src/network/gossip/service.rs:507]: Error type plx.

File: finality-aleph/src/party/manager/aggregator.rs [hunk 7]
```diff
@@ -91,7 +93,7 @@ async fn run_aggregator<B, C, CN, LN, JS>(
     session_boundaries: &SessionBoundaries,
     mut metrics: AllBlockMetrics,
     mut exit_rx: oneshot::Receiver<()>,
-) -> Result<(), ()>
+) -> Result<(), String>
```
Comment (timorleph) [inline finality-aleph/src/party/manager/aggregator.rs:96]: Error type plx.

File: finality-aleph/src/party/manager/aggregator.rs [hunk 8]
```diff
@@ -209,6 +208,13 @@ where
                 exit,
             )
             .await;
+            let result = match result {
+                Ok(_) => Ok(()),
+                Err(err) => {
```
Comment (timorleph) [inline finality-aleph/src/party/manager/aggregator.rs:213]: If you have a real error then better return it rather than masking it here.

File: finality-aleph/src/sync/service.rs [hunk 9]
```diff
@@ -730,54 +731,61 @@ where
     }
 
     /// Stay synchronized.
-    pub async fn run(mut self) {
+    pub async fn run(mut self) -> Result<Bottom, String> {
```
Comment (timorleph) [inline finality-aleph/src/sync/service.rs:734]: Error types plx, here especially important since you get wrappers much more naturally than doing a `format!` in a `map_err`.

File: primitives/src/lib.rs [hunk 10]
```diff
@@ -61,6 +61,12 @@ pub type Nonce = u32;
 /// A hash of some data used by the chain.
 pub type Hash = sp_core::H256;
 
+/// Can be used to represent an infinite computation. One can use it together
+/// with [Result<Bottom, T>](Result) to indicate that a computation can finish
+/// only when some error occurs, otherwise it will spin forever
+/// (compiler/type-checker will enforce this property).
```
Comment (timorleph) [inline primitives/src/lib.rs:67]: This is kinda cool (⊥ is my favourite type in general), but I'm not convinced it's worth the effort if it's not a native type. I doubt everyone will start using it now, so we will only have this in a couple semi-random places, where it'll confuse any readers.

If we want to adopt this, then I would suggest making an ADR for that and actually requiring people to put ⊥s everywhere where there should be a never ending computation.

File: finality-aleph/src/data_io/data_provider.rs (hunk 3 above)
Comment (timorleph) [reply finality-aleph/src/data_io/data_provider.rs:296]: Hm, I don't want to keep `Bottom` as a type in general (see another comment in this review), but wouldn't an empty `match` be technically exhaustive, and thus solve this issue? That is:
```
impossible_result = self.main_loop() => match impossible_result {},
```

**Review by timorleph — CHANGES_REQUESTED** [review]
Cool, although I'm worried that the changes are not consistent between loops – it would be great if you could make them so.

File: clique/src/service.rs [hunk 11]
```diff
@@ -205,12 +208,94 @@ where
...
+                if let Some(address) = self.peer_address(&public_key) {
+                    self.spawn_new_outgoing(public_key, address, result_for_parent.clone());
+                }
+            }
+        }
+    }
+
     /// Run the service until a signal from exit.
-    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) -> Result<(), String> {
```
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply clique/src/service.rs:295]: I am going to name it `GenericStringError` ;P

File: primitives/src/lib.rs (hunk 10 above)
Comment (DamianStraszak) [reply primitives/src/lib.rs:67]: Can't we use https://doc.rust-lang.org/std/primitive.never.html ?

Comment (fixxxedpoint) [reply primitives/src/lib.rs:67]: https://doc.rust-lang.org/reference/types/never.html

File: clique/src/service.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply clique/src/service.rs:295]: I have mixed feelings with this. Only thing we can do with these errors is logging, we do not handle them in any way. I can introduce a new error that will be just a wrapper for `String`. Then I will have to write `Into/From` for it for `&str` conversions for the `?` operator, etc. It will be pure boilerplate. For these errors we don't even care about backtraces. Perhaps just use `Error` from anyhow (new dependency)?

File: finality-aleph/src/block/substrate/status_notifier.rs (hunk 4 above)
Comment (fixxxedpoint) [reply finality-aleph/src/block/substrate/status_notifier.rs:70]: https://docs.rs/futures/latest/futures/macro.select.html

File: finality-aleph/src/metrics/chain_state.rs [hunk 12]
```diff
@@ -176,21 +174,29 @@ pub async fn run_chain_state_metrics<
...
-            maybe_block = finality_notifications.next() => match maybe_block {
+            maybe_block = finality_notifications.next(), if !finality_notifications.is_terminated() => match maybe_block {
                 Some(block) => handle_block_finalized(block, &metrics),
                 None => warn!(
                     target: LOG_TARGET,
                     "Finality notification stream ended unexpectedly"
                 ),
             },
-            maybe_transaction = transaction_pool_info_provider.next_transaction() => {
+            maybe_transaction = transaction_pool_info_provider.next_transaction(), if !transaction_pool_info_provider_terminated => {
```
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply finality-aleph/src/metrics/chain_state.rs:184]: Short version: this one doesn't `early` exit on an empty channel. Why? No idea. @ggawryal ? I will just rewrite it into early-exit version similar to clique.

File: primitives/src/lib.rs (hunk 10 above)
Comment (DamianStraszak) [reply primitives/src/lib.rs:67]: Right -- the question is then if we need `Bottom` at all

File: finality-aleph/src/data_io/data_provider.rs (hunk 3 above)
Comment (fixxxedpoint) [reply finality-aleph/src/data_io/data_provider.rs:296]: What about using `futures::future::select` ?

File: finality-aleph/src/party/manager/aggregator.rs (hunk 8 above)
Comment (fixxxedpoint) [reply finality-aleph/src/party/manager/aggregator.rs:213]: Do you have something more specific on your mind, cuz I can't just return some error here without refactoring Task/Subtask, etc.

File: primitives/src/lib.rs (hunk 10 above)
Comment (fixxxedpoint) [reply primitives/src/lib.rs:67]: I am ok with introducing it just in few random places (why do we need to use it everywhere immediately?), is more descriptive for infinite loops/services than `Result<(), _>`, it gives some free additional compiler checks, it doesn't require any additional maintenance (no methods, converters etc.), and can be switched to `!` when it become stable (are we going to use it then or pretend it's not there?). ADR for it is an overkill, since it will be a rust's feature anyway. I can change it's name for something more descriptive. It was interesting to see that you can do it in rust. Anyway, should I remove it?

File: finality-aleph/src/metrics/chain_state.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ggawryal) [reply finality-aleph/src/metrics/chain_state.rs:184]: Yes, we can rewrite this in clique style (had no particular reason why in mind, when I was writing that :))

File: finality-aleph/src/data_io/data_provider.rs [hunk 13]
```diff
@@ -280,25 +280,23 @@ where
...
+            _ = exit => debug!(target: "aleph-data-store", "Task for refreshing best chain received exit signal. Terminating."),
```
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply finality-aleph/src/data_io/data_provider.rs]: changed

File: finality-aleph/src/data_io/data_provider.rs (hunk 3 above)
Comment (ggawryal) [reply finality-aleph/src/data_io/data_provider.rs]: Looks good

File: clique/src/service.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (DamianStraszak) [reply clique/src/service.rs:295]: I think defining the errors using `thiserror` could make sense. This way you have proper types implementing `Error` and in the future you can handle this errors easily. But let's wait for @timorleph

File: finality-aleph/src/data_io/data_store.rs [hunk 14]
```diff
@@ -237,26 +238,36 @@ where
...
         let mut maintenance_clock = Delay::new(self.config.periodic_maintenance_interval);
         let mut import_stream = self.client.import_notification_stream();
+        if import_stream.is_terminated() {
+            return Err("Blocks import notification stream was closed.")?;
```
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply finality-aleph/src/data_io/data_store.rs]: changed

File: finality-aleph/src/data_io/data_store.rs [hunk 15]
```diff
@@ -237,26 +238,36 @@ where
...
+        }
         let mut finality_stream = self.client.finality_notification_stream();
+        if finality_stream.is_terminated() {
+            return Err("Finalized-blocks notification stream was closed.")?;
```
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply finality-aleph/src/data_io/data_store.rs]: changed

File: finality-aleph/src/sync/service.rs (hunk 1 above)
Comment (fixxxedpoint) [reply finality-aleph/src/sync/service.rs]: changed

File: finality-aleph/src/sync/service.rs (hunk 2 above)
Comment (fixxxedpoint) [reply finality-aleph/src/sync/service.rs]: changed

File: clique/src/service.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (timorleph) [reply clique/src/service.rs]: Just normal enum errors are much nicer, lets not add another layer of problems in case we ever manage to get to implement actual tests for these things (in which case "stringy" errors would be extremely annoying). If you are worried about the boilerplate of defining all the `Display`s and whatnot then @DamianStraszak 's suggestion of `thiserror` is reasonable – tbh we probably should've been using it all along (at the very least in node, ABFT might have some argument about minimizing deps).

File: finality-aleph/src/party/manager/aggregator.rs (hunk 8 above)
Comment (timorleph) [reply finality-aleph/src/party/manager/aggregator.rs:234]: Ugh, they probably need a refactor then. Don't necessarily do this in this PR, but please create a task for it.

File: primitives/src/lib.rs (hunk 10 above)
Comment (timorleph) [reply primitives/src/lib.rs:67]: The problem with doing it in just a couple random places is that there is additional mental load when trying to figure out why some places have it and others don't – and when the answer is "because it was done in one specific PR" it's not even something that can really be guessed. There is value in consistency. This would be slightly less of a problem if we used native Rust functionalities, but still somewhat, we don't use all the Rust functionalities that exist and would technically improve compiler checks. See for example the infamous lifetimed keychain reference in ABFT.

That being said, I see a couple reasonable approaches to this:
1. Leave this PR in its current shape, but create an ADR saying that we use `Bottom` for all functions where it's applicable and switch to `!` as soon as it's stable. This way we get temporarily confusing code, that will eventually get homogenized, which is fine. The ADR can even be very short, the argument for using this type is pretty straightforward.
2. Remove `Bottom`, but create an ADR for using some type like it eventually, probably once `!` is stabilized.
3. Remove `Bottom` and create an ADR explicitly forbidding the use of `!` to avoid confusion.
4. Remove `Bottom` and do nothing.
This is roughly in my order of preference. Note that 4 is essentially doing 3 implicitly, as there will be a discussion like this every time someone tries to re-add `Bottom`. For 2 & 3 you don't have to create any ADR immediately, it's just something that would eventually have to be done. Also, even if we go for 4 and then accept a PR with `Bottom` (probably in the form of a stabilized `!`), we would still have to at least plan an effort to homogenize the code.

tl;dr Either leave `Bottom` in and create an ADR, or remove bottom and optionally create a task for writing an ADR.

File: finality-aleph/src/metrics/chain_state.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (fixxxedpoint) [reply finality-aleph/src/metrics/chain_state.rs]: done

File: finality-aleph/src/network/gossip/service.rs (hunk 6 above)
Comment (fixxxedpoint) [reply finality-aleph/src/network/gossip/service.rs]: done

File: finality-aleph/src/party/manager/aggregator.rs (hunk 7 above)
Comment (fixxxedpoint) [reply finality-aleph/src/party/manager/aggregator.rs]: done

File: finality-aleph/src/party/manager/aggregator.rs (hunk 8 above)
Comment (fixxxedpoint) [reply finality-aleph/src/party/manager/aggregator.rs:242]: done (added task)

File: finality-aleph/src/sync/service.rs (hunk 9 above)
Comment (fixxxedpoint) [reply finality-aleph/src/sync/service.rs]: done

File: clique/src/service.rs [hunk 16]
```diff
@@ -205,12 +247,99 @@ where
...
+        }
+    }
+
     /// Run the service until a signal from exit.
-    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+    pub async fn run(
+        mut self,
+        mut exit: oneshot::Receiver<()>,
+    ) -> Result<(), CliqueNetworkServiceError> {
+        use CliqueNetworkServiceError as Error;
```
Comment (timorleph) [inline clique/src/service.rs:338]: Tbh I think you can just call it `Error` originally, you only have one in this module. :D

File: clique/src/service.rs [hunk 17]
```diff
@@ -85,6 +92,41 @@ pub trait SpawnHandleT {
...
+            }
+            CliqueNetworkServiceError::ConnectionWorker => {
+                write!(f, "`worker_results` channel was closed.")
+            }
+        }
+    }
+}
+
+impl CliqueNetworkServiceError {
+    fn commands_stream_closed() -> Self {
```
Comment (timorleph) [inline clique/src/service.rs:117]: These constructors don't really seem necessary, no?

File: finality-aleph/src/data_io/data_store.rs [hunk 18]
```diff
@@ -105,6 +107,43 @@ impl Default for DataStoreConfig {
...
+                write!(f, "Finalized block import notification stream was closed.")
+            }
+            DataStoreError::NetworkMessagesTerminated => {
+                write!(f, "Stream with network messages was closed.")
+            }
+        }
+    }
+}
+
+impl DataStoreError {
```
Comment (timorleph) [inline finality-aleph/src/data_io/data_store.rs:133]: Same comments as in with the other error – name could just be `Error` and constructors are not necessary.

File: finality-aleph/src/data_io/legacy/data_store.rs [hunk 19]
```diff
@@ -106,6 +108,43 @@ impl Default for DataStoreConfig {
...
+                write!(f, "Finalized block import notification stream was closed.")
+            }
+            DataStoreError::NetworkMessagesTerminated => {
+                write!(f, "Stream with network messages was closed.")
+            }
+        }
+    }
+}
+
+impl DataStoreError {
```
Comment (timorleph) [inline finality-aleph/src/data_io/legacy/data_store.rs:134]: Same.

File: finality-aleph/src/metrics/chain_state.rs [hunk 20]
```diff
@@ -135,59 +186,57 @@ pub async fn run_chain_state_metrics<
...
+    if finality_notifications.is_terminated() {
+        return Err(Error::finalized_blocks_stream_closed());
     }
 
     let metrics = match ChainStateMetrics::new(registry) {
         Ok(metrics) => metrics,
         Err(e) => {
             warn!(target: LOG_TARGET, "Failed to create metrics: {e}.");
-            return;
+            return Err(Error::unable_to_create_metrics(e));
```
Comment (timorleph) [inline finality-aleph/src/metrics/chain_state.rs:210]: You can probably `map_err()?` here, instead of this whole `match`. The log ain't that useful and should be duplicated by error handling outside anyway.

File: finality-aleph/src/metrics/chain_state.rs [hunk 21]
```diff
@@ -31,6 +31,57 @@ const TRANSACTION_CACHE_SIZE: usize = 100_000;
...
+                write!(f, "Finality notification stream ended unexpectedly.")
+            }
+            ChainStateMetricsError::TransactionStreamClosed => {
+                write!(f, "Transaction stream ended unexpectedly.")
+            }
+        }
+    }
+}
+
+impl ChainStateMetricsError {
```
Comment (timorleph) [inline finality-aleph/src/metrics/chain_state.rs:63]: Same.

File: finality-aleph/src/network/gossip/service.rs [hunk 22]
```diff
@@ -77,6 +78,49 @@ impl Display for Error {
...
+                write!(f, "Block sync user message stream ended.")
+            }
+            GossipServiceError::UnableToForwardMessageToUser => {
+                write!(f, "Cannot forward messages to user.")
+            }
+        }
+    }
+}
+
+impl GossipServiceError {
```
Comment (timorleph) [inline finality-aleph/src/network/gossip/service.rs:106]: Same.

File: finality-aleph/src/party/manager/aggregator.rs [hunk 23]
```diff
@@ -27,6 +29,33 @@ use crate::{
...
+        match self {
+            AggregatorError::MultisignaturesStreamTerminated => {
+                write!(f, "The stream of multisigned hashes has ended.")
+            }
+            AggregatorError::UnableToProcessHash => write!(f, "Error while processing a hash."),
+        }
+    }
+}
+
+impl AggregatorError {
```
Comment (timorleph) [inline finality-aleph/src/party/manager/aggregator.rs:49]: Same.

File: finality-aleph/src/sync/service.rs [hunk 24]
```diff
@@ -81,6 +81,70 @@ where
...
+                    f,
+                    "Channel with legacy internal block request from user closed."
+                )
+            }
+            SyncServiceError::CreatorChannelClosed => write!(f, "Channel with own blocks closed."),
+        }
+    }
+}
+
+impl<NetworkError, ChainEventError> SyncServiceError<NetworkError, ChainEventError> {
```
Comment (timorleph) [inline finality-aleph/src/sync/service.rs:122]: Same.

File: clique/src/service.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (timorleph) [reply clique/src/service.rs]: Well, I guess we are not addint `thiserror` then, oh well, maybe next time or when we have some time for a refactor...

**Review by timorleph — APPROVED** [review]
Cool, just a couple small nits.

File: clique/src/service.rs (hunk 17 above)
Comment (fixxxedpoint) [reply clique/src/service.rs:117]: :D

Comment (timorleph) [reply clique/src/service.rs:117]: I mean, you can just use `enum` variants as constructors as we do in all other code, right? <_<"

Comment (fixxxedpoint) [reply clique/src/service.rs:117]: I thought the goal was to make it future-proof. This way we can add more details to them later more easily (e.g. back-traces). Just kidding ;)

File: finality-aleph/src/metrics/chain_state.rs (hunk 20 above)
Comment (fixxxedpoint) [reply finality-aleph/src/metrics/chain_state.rs]: done

File: finality-aleph/src/metrics/chain_state.rs (hunk 21 above)
Comment (fixxxedpoint) [reply finality-aleph/src/metrics/chain_state.rs]: done

File: finality-aleph/src/network/gossip/service.rs (hunk 22 above)
Comment (fixxxedpoint) [reply finality-aleph/src/network/gossip/service.rs]: done

File: finality-aleph/src/sync/service.rs (hunk 24 above)
Comment (fixxxedpoint) [reply finality-aleph/src/sync/service.rs]: done

File: finality-aleph/src/network/gossip/service.rs (hunk 22 above)
Comment (timorleph) [reply finality-aleph/src/network/gossip/service.rs]: This and the other I left unresolved were not renamed to just `Error` – is there a reason for that?

Comment (fixxxedpoint) [reply finality-aleph/src/network/gossip/service.rs]: `ServiceInterface` uses `Error`.

File: finality-aleph/src/sync/service.rs (hunk 24 above)
Comment (fixxxedpoint) [reply finality-aleph/src/sync/service.rs]: done
